### PR TITLE
Improve TerriaViewer and specs

### PR DESCRIPTION
**Pull request by tephenavies**
### What this PR does

- Move async import of Leaflet and Cesium outside of TerriaViewer
- Refactor TerriaViewer's async viewer loading to expose a promise viewerLoadPromise and to track load promises outside the class.
- Add viewer cleanup code to many specs
- Fix temperamental tests - added an undefined check to terrain sampling code that would sometimes throw exceptions in tests when you mouse over a div being used in automated testing.

### Test me

When you run specs from http://localhost:3002/SpecRunner.html?spec= you no longer get a long page of used cesium divs after specs finish.

### Checklist

- [x] There are unit tests to verify my changes are correct or unit tests aren't applicable (if so, write quick reason why unit tests don't exist)
- [x] No doc changes needed.
- [x] I've updated CHANGES.md with what I changed.
- [x] I've provided instructions in the PR description on how to test this PR.

### CODE REVIEW

File: lib/Models/NoViewer.ts [hunk 1]
```diff
@@ -11,16 +11,14 @@ import CameraView from "./CameraView";
...
 
 class NoViewer extends GlobeOrMap {
   readonly type = "none";
   readonly terria: Terria;
   readonly canShowSplitter = false;
   private _currentView: CameraView = new CameraView(Rectangle.MAX_VALUE);
-  readonly dataAttributions = observable([]);
 
-  constructor(terriaViewer: TerriaViewer) {
+  constructor(terriaViewer: TerriaViewer, _container?: string | HTMLElement) {
```
Comment (tephenavies) [inline lib/Models/NoViewer.ts:21]: Match constructor parameters of Cesium and Leaflet

File: lib/ReactViewModels/MouseCoords.ts [hunk 2]
```diff
@@ -78,7 +78,9 @@ export default class MouseCoords {
     const camera = scene.camera;
     const pickRay = camera.getPickRay(position);
     const globe = scene.globe;
-    const pickedTriangle = (<any>globe).pickTriangle(pickRay, scene);
+    const pickedTriangle = isDefined(pickRay)
+      ? (<any>globe).pickTriangle(pickRay, scene)
+      : undefined;
```
Comment (tephenavies) [inline lib/ReactViewModels/MouseCoords.ts:98]: Fix source of intermittent errors when running tests in an interactive window

File: lib/ViewModels/TerriaViewer.ts [hunk 3]
```diff
@@ -47,23 +61,21 @@ export default class TerriaViewer {
...
+                : baseMap.uniqueId) ?? "Unknown item"
           }
-        });
-      } else {
-        runInAction(() => (this._baseMap = baseMap));
-      }
+        }
+      });
+    } else {
+      runInAction(() => (this._baseMap = baseMap));
```
Comment (tephenavies) [inline lib/ViewModels/TerriaViewer.ts:78]: Remove unnecessary `if (baseMap) {`

File: test/Models/CesiumSpec.ts [hunk 4]
```diff
@@ -127,68 +127,75 @@ describeIfSupported("Cesium Model", function () {
...
+      await runLater(() => {});
+
+      // Test that we have removed the correct items
+      expect(dataSourceNames()).toEqual(["ds3"]);
+      expect(tilesetUrls()).toEqual(["prim3"]);
+      expect(imageryProviderUrls()).toEqual(["img3"]);
+    } finally {
+      cesium2.destroy();
+      document.body.removeChild(container2);
+    }
```
Comment (tephenavies) [inline test/Models/CesiumSpec.ts]: Change variables names to not clash, and wrap in try, finally to ensure cesium & container is cleaned up

File: test/Models/UserDrawingSpec.ts [hunk 5]
```diff
@@ -6,39 +6,42 @@ import Ellipsoid from "terriajs-cesium/Source/Core/Ellipsoid";
...
-      .catch(done.fail);
+    const cesium = terria.cesium;
+    expect(cesium).toBeDefined();
+    if (cesium) {
+      expect(cesium.cesiumWidget.canvas.style.cursor).toEqual("");
+      userDrawing.enterDrawMode();
+      expect(cesium.cesiumWidget.canvas.style.cursor).toEqual("crosshair");
+      (<any>userDrawing).cleanUp();
+      expect(cesium.cesiumWidget.canvas.style.cursor).toEqual("auto");
+    }
```
Comment (tephenavies) [inline test/Models/UserDrawingSpec.ts:44]: Use beforeEach and afterEach to ensure everything is cleaned up.

File: lib/ViewModels/TerriaViewer.ts [hunk 6]
```diff
@@ -21,8 +21,19 @@ import NoViewer from "../Models/NoViewer";
...
+  () =>
+    fromPromise(import("../Models/Leaflet").then((Leaflet) => Leaflet.default)),
+  { keepAlive: true }
+);
+
+const cesiumFromPromise = computed(
+  () =>
+    fromPromise(import("../Models/Cesium").then((Cesium) => Cesium.default)),
+  { keepAlive: true }
+);
```
Comment (tephenavies) [inline lib/ViewModels/TerriaViewer.ts:36]: Moved tv._cesiumPromise & tv._leafletPromise out to module scope. These are still only loaded when needed (although at the moment in `main` there is no separate Cesium bundle).

**Review by na9da — APPROVED** [review]
That's some good refactoring @steve9164 :rocket:

**Pull request merged**